### DISCUSSION TITLE
How would I integrate something like feather-icons without errors?

### DISCUSSION THREAD

**Discussion by lyssieth** [leptos-rs/leptos#719] [Q&A] [answered]
Currently I have it set up as follows:

A `Feather` component, that just becomes this:
```html
<script src="https://unpkg.com/feather-icons" />
<script defer>
    feather.replace();
</script>
```

And `Icon` components, which are again quite simple.
```rs
#[component]
pub fn Icon(cx: Scope, #[prop(into)] name: String) -> impl IntoView {
    view! {
        cx,
        <i data-feather=name></i>
    }
}
```

After the `feather.replace` call from the script goes through, the `<i>` tags get replaced with svgs, which is how Feather works as far as I could tell..

And the Icon component, while it *does* render, spews out errors into the console.
![image](https://user-images.githubusercontent.com/1861988/226712258-492864bb-6c10-417b-8a82-2544e1b6aca3.png)
[Full error as text](https://gist.github.com/lyssieth/1e9a2dbf5ef8b8353df956473c3e4794)

**Comment by gbj** [comment, marked as answer]
(I often link to GitHub discussions when other people need help, so I've written kind of a long answer here. Feel free to skip to the end but I think reading the first will help you when you're thinking about how thi framework, or honestly most similar JS or WASM frameworks, works overall.)

## The Problem

So let's think through the way your application runs:
- you make a request to the server
- the server renders your app to HTML, including
  1. a `<head>` that tells the browser to start loading the WASM version of your app
  2. a `<body>` that includes your app, rendered to HTML, including
    i. the `<script>` tags from the Feather component
    ii. the `<i>` tag from the Icon component
- the browser does three things simultaneously
  - the browser displays the HTML, which includes (I assume?) blank spaces where the icons will load
  - the browser starts loading the Feather script
    - when this resolves, it replaces all the `<i data-feather>` elements with `<svg>` elements 
  - the browser also starts loading your WASM
    - when this resolves, it walks over your app, adding any necessary event listeners or reactivity

If the WASM loads before the Feather JS, it's fine (for now). It says, "Okay, there's an `<i>`
 tag... nothing to add here for reactivity" and moves on.

If the Feather JS loads before the WASM, it swaps out all the `<i data-feather>` elements. But now when the WASM finishes loading and hydrates it says, "Wait, what? You told me this is an `<i>` but it's an `<svg>`" and bails out. As the error text says, this is a discrepancy between SSR (server side rendering) and CSR (client-side rendering).

Now, you could say — Well, why don't we just ignore any mismatches like this? And there are cases in which you'd be right.

However, on the first client-side navigation to another route with an icon, or the first time you try to create the Icon element on the client side, you'd come back and report that the icons don't load. Which makes sense, because the Feather JS loads once on page load and calls the `replace()` function once, on page load. It can't do anything to update `<i data-feather>` that you create later.

In my experience, these "just load some JS from a CDN and run it" approaches really only work with static HTML sites/multi-page apps that are just loading HTML from the server on every navigation.

## Two Solutions
1. I add a `<NoHydration/>` component to the framework which just skips over its children during hydration. This is possible, and on the roadmap for the future, but doesn't exist yet. This would not really solve the problem of any Icon components created after initial load.
2. Since Feather is basically just a set of SVG icons, you can design an Icon component that loads those SVGs, perhaps as `<object>` elements, either locally or from a CDN. This will actually load much faster for your users, because it doesn't have to hit up the CDN for a JS blob including all the components, and it has the advantage of being good semantic HTML instead of `<i>` tags.

Option 2 is the clear winner from my perspective. It looks something like this

## `<Icon/>` 2.0
```rust
#[component]
pub fn Icon(cx: Scope, #[prop(into)] name: String) -> impl IntoView {
    let path = format!("/path-to-images/{name}.svg";
    view! {
        cx,
        <object data=path type="image/svg+xml"></object>
    }
}
```

**Reply by lyssieth** [reply]
Thank you!

I went with option 3, which is switching to Google's Material Icons instead. But... both of these would've been solutions as well, yes.

Truthfully, I haven't worked with frameworks like this before, as my past experiences are very static in comparison. So it's good to expand my knowledge as well.
